doc/rst/language/spec/procedures.rst

```diff
@@ -532,114 +630,142 @@ model.
+The following example shows such modification by means other than the
+``const ref`` formal. This program would have implementation-dependent
+behavior if ``c`` used ``const`` intent instead of ``const ref`` (see
+:ref:`The_Const_Intent`).
+
+  *Example (const-ref-aliasing.chpl)*.
+
+   .. code-block:: chapel
+
+      var a: int = 0;
+      proc f(ref b, const ref c) {
+        writeln(c); // outputs 0
+        a = 1;
+        writeln(c); // outputs 1, because 'c' refers to 'a'
+        b = 2;      // modifies 'a' because 'b' refers to 'a'
+        writeln(c); // outputs 2, because 'c' refers to 'a'
+      }
+      f(a, a);
+
+   .. BLOCK-test-chapeloutput
+
+      0
+      1
+      2
 
 .. _The_Const_Intent:
 
 The Const Intent
-^^^^^^^^^^^^^^^^
+~~~~~~~~~~~~~~~~
+
+The ``const`` intent is broadly similar to the ``const ref`` intent;
+however it allows the implementation more freedom to optimize.  Generally
+speaking, the compiler will implement a ``const`` intent formal by
+choosing between ``const in`` and ``const ref`` in an
+implementation-defined manner. As such, it is appropriate when the actual
+value will not be modified anywhere within the function and additionally
+the actual value will not be modified through other means than the
+``const ref`` formal. See the example in :ref:`The_Const_Ref_Intent` for
+an example of modification through other means.
+
+The ``const`` intent indicates that the function will not and cannot
+modify the formal argument within its dynamic scope.
+
+Unlike ``const in`` / ``in``, the ``const`` intent cannot be relied upon
+to transfer a value. For record types, passing to a ``const`` intent will
+never cause copy initialization.
+
+For synchronization types ``sync`` and ``atomic``, the ``const``
+intent is ``const ref`` and the remainder of this section does not apply.
+
+Otherwise, the ``const`` intent communicates to the compiler that it
+can make several assumptions to aid optimization within the function using
+a ``const`` formal:
+
+ * the value referred to by such a formal argument will not be modified
+   while the function is running
+
+ * similarly, if the formal is a ``domain``, the domain will not be
+   while the function is running
 
-The ``const`` intent specifies that the function will not and cannot
-modify the formal argument within its dynamic scope. Whether ``const``
-is interpreted as ``const in`` or ``const ref`` intent depends on the
-argument type.  Generally, small values, such as scalar types, will be
-passed by ``const in``; while larger values, such as domains and
-arrays, will be passed by ``const ref`` intent.
+ * if the formal is an array, value of the domain that the array is
+   declared over will not be modified while the function is running
 
-For the following types, the ``const`` intent means ``const in``:
+ * if the formal is an ``owned`` or ``shared``, the ``owned`` or
+   ``shared`` will not be modified to point to a different object while
+   the function is running.
```

Comment:
"the owned or shared" sounds vague, I suggest "the actual will not be modified". "to point to a different object" --> also "or nil".